zsh doesn't build with 6 series ncurses

wspear opened the pull request:
See: https://github.com/spack/spack-packages/issues/1024

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

wspear committed:
zsh doesn't build with 6 series ncurses

See: https://github.com/spack/spack-packages/issues/1024

wspear committed:
Merge branch 'develop' into patch-1

alecbcs commented:
We should probably investigate this further as neither [nix](https://github.com/NixOS/nixpkgs/blob/nixos-25.05/pkgs/by-name/zs/zsh/package.nix) or [brew](https://github.com/Homebrew/homebrew-core/blob/7606217f5605c3db14cf0934e0ea5d8b576a6eed/Formula/z/zsh.rb) limit the version of ncurses used in their builds.

michaelkuhn commented:
I'm guessing https://github.com/zsh-users/zsh/commit/4c89849c98172c951a9def3690e8647dae76308f.patch is the patch we need (used in the Nix and Fedora recipes).

alecbcs commented:
@wspear can you try relaxing dependency limit on `ncurses@5` and add the linked patch and see if that also fixes the build issue for you?

alecbcs commented:
@wspear do you have time to try the linked patch and update the PR?

wspear commented:
The patch operation failed and I'm not sure why. I downloaded it to the zsh directory and added patch("4c89849c98172c951a9def3690e8647dae76308f.patch") on line 47 of package.py and got:

```
==> Installing zsh-5.9-gmazlahrrlb7gxvdyqbnr2t7lmmtq5ng [8/8]
==> Using cached archive: /home/wspear/bin/SPACK/spack/var/spack/cache/_source-cache/archive/9b/9b8d1ecedd5b5e81fbf1918e876752a7dd948e05c1a0dba10ab863842d45acd5.tar.xz
1 out of 1 hunk FAILED -- saving rejects to file ChangeLog.rej
==> Patch /home/wspear/.spack/package_repos/fncqgg4/repos/spack_repo/builtin/packages/zsh/4c89849c98172c951a9def3690e8647dae76308f.patch failed.
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/patch' '-s' '-p' '1' '-i' '/home/wspear/.spack/package_repos/fncqgg4/repos/spack_repo/builtin/packages/zsh/4c89849c98172c951a9def3690e8647dae76308f.patch' '-d' '.'
```

michaelkuhn commented:
This happens due to the `ChangeLog` changes, just remove that hunk.

wspear commented:
That worked and zsh built. But I also had to remove the commit id (changed the filename and removed the line internally, not sure which worked) to prevent spack from rejecting it due to the hash difference. Is there a best practice for committing someone else's edited patch?

wspear committed:
Add patch to avoid pointer type issues

wspear committed:
Use patch for pointer types

Patch only compatible with 5.6.2 and up. Earlier versions fail to build for apparently unrelated reasons.

wspear commented:
Thanks for tracking that patch down. This pr works for me.

eugeneswalker commented:
Any objections to merging this? Can we proceed @michaelkuhn @alecbcs ?

alecbcs reviewed (approved):
Looks good to me. Thanks @wspear for digging into this patch.
